Pagination support (NOT READY TO MERGE)

furuholm opened the pull request:
This PR is a partial solution for #100. Only project pagination has been implemented. The reason for submitting a partial solution is to enable discussions before implementing pagination for all resource types.

Notes:
- I see a number of options for obfuscating page tokens:
  1. Don't
  2. Encode with e.g. base64
  3. Encrypt the tokens using a key that is only known to the server
  
  The memstore backend uses option 1in this PR (for simplicity).
  The postgres backend uses option 2 in this PR.
  Clair uses option 3.
  I started out with option 3 but figured I would keep it simple until I got some feedback. What is the preferred option   for the postgres backend? Is there any best practice at google? 

- I added a ListOptions struct to keep the number of parameters down. Not sure if this was an improvement or not. Feedback welcome.

- We might have to add new db indexes if the upcomming filtering functionality will support ordering.

R2wenD2 commented on server-go/storage.go:
```diff
@@ -19,6 +19,16 @@ import (
 	opspb "google.golang.org/genproto/googleapis/longrunning"
 )
 
+type ListOptions struct {
```
@vtsao Do you have thoughts on this struct? I have no issues with it.

vtsao replied:
I would just inline these fields as args to the method, it makes it easier to read and I don't think we expect this to grow anytime soon.

furuholm replied:
Fixed in 724e2e3

R2wenD2 commented on samples/server/go-server/api/server/storage/memstore.go:
```diff
@@ -42,53 +43,62 @@ func NewMemStore() server.Storager {
 }
 
 // GetProject returns the project with the given pID from the mem store
 func (m *memStore) GetProject(pID string) (*pb.Project, error) {
```
Perhaps we should keep the map for efficient gets and then just use the sorted keys of the map for the List method? What do you think? 

This isn't a huge deal since its in memory and the size of things shouldn't be too large.

furuholm replied:
Agree. Switching to lists turned out to be more intrusive than I first thought so going back to a map is better for several reasons.

furuholm replied:
OMG. So much ugly code was introduced by this change. What was I thinking?!? Good thing you reviewed this.😄 

363992e reverts back to storing projects in a map.

R2wenD2 commented on samples/server/go-server/api/server/storage/memstore.go:
```diff
@@ -42,53 +43,62 @@ func NewMemStore() server.Storager {
 		occurrencesByID: map[string]*pb.Occurrence{},
 		notesByID:       map[string]*pb.Note{},
 		opsByID:         map[string]*opspb.Operation{},
```
What are your plans for pagination on these? We can get all of the values for a project and sort them, I guess.

furuholm replied:
Will add pagination for them as well. Just wanted the projects to be done first. And yes, if we sort them we should be fine.

jzelinskie commented:
You should probably use [fernet](https://godoc.org/github.com/fernet/fernet-go) rather than base64 to avoid leaking primary keys from the database. This is what clair uses.

R2wenD2 commented:
Sorry - I wasn't clear, I meant an encrypted base64 encoded string. fernet also looks fine.

Agree that we need to support migrations, we'll eventually get to that point :)

furuholm commented:
Ok. I'll add page token encryption.

R2wenD2 commented:
@furuholm is this ready for another look?

furuholm commented:
@R2wenD2 I haven't got around to it. I will try to get some time to do it during this week. Do you have other work that is being held up by this?

R2wenD2 commented:
no blockers on my part. just wanted to make sure you weren't waiting on me for a review :)

furuholm commented:
Tokens are now encrypted using fernet-go.

Will keep this PR open until we are satisfied with project pagination. Will create a new PR that adds pagination for the other entity types.

vtsao commented on samples/server/go-server/api/server/storage/pgsqlstore.go:
```diff
@@ -126,22 +130,36 @@ func (pg *pgSQLStore) GetProject(pID string) (*pb.Project, error) {
-// ListProjects returns the project id for all projects from the store
-func (pg *pgSQLStore) ListProjects(filters string) ([]*pb.Project, error) {
-	rows, err := pg.DB.Query(listProjects)
+var paginationKey = "XxoPtCUzrUv4JV5dS+yQ+MdW7yLEJnRMwigVY/bpgtQ="
```
Is this meant to be here, shouldn't this be read from the config?

furuholm replied:
You are right. Forgot to remove this. Fixed in 5108566.

vtsao commented on samples/server/go-server/api/server/storage/pgsqlstore.go:
```diff
@@ -126,22 +130,36 @@ func (pg *pgSQLStore) GetProject(pID string) (*pb.Project, error) {
+	}
 	if err != nil {
-		return nil, status.Error(codes.Unknown, fmt.Sprintf("Failed to list Projects from database"))
+		return nil, "", status.Error(codes.Unknown, "Failed to list Projects from database")
```
codes.Internal is probably better here and below

furuholm replied:
Agree. Fixed in 2c060cf

vtsao commented on samples/server/go-server/api/server/storage/pgsqlstore.go:
```diff
@@ -472,3 +490,23 @@ func (pg *pgSQLStore) ListOperations(pID, filters string) ([]*opspb.Operation, e
+
+// Encrypt int64 using provided key
+func encryptInt64(v int64, key string) (string, error) {
+	k, _ := fernet.DecodeKey(key)
```
You should check the error here, same for decrypt.

vtsao commented on samples/server/go-server/api/server/storage/pgsqlstore.go:
```diff
@@ -472,3 +490,23 @@ func (pg *pgSQLStore) ListOperations(pID, filters string) ([]*opspb.Operation, e
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
```
we should encode this in base64

furuholm replied:
The token returned from fernet-go is already base64 encoded.

furuholm commented:
@vtsao: I've pushed fixes that addresses your comments, but I didn't rebase the changes so there are a a bunch of other things that has been added to this PR that really should be put in a new PR. You can ignore them for now if you want to.

furuholm commented:
I've created a new PR #115 where I have squashed and rebased the commits.

(I've pushed all of the commits that was squashed into #115 to this PR if anyone prefers reviewing individual changes.)
